### ISSUE TITLE
Add support for 'missing' RPCs

### ISSUE THREAD

**Issue by gavanderhoorn** [gavanderhoorn/comet_rpc#1]
`comet_rpc` currently only supports a subset of the RPCs `COMET` offers.

Implementing the missing ones should not be too difficult:

 1. figure out their names + arguments (using Wireshark fi, or by looking at the Javascript)
 2. their behaviour
 3. returned response document (JSON)
 4. add class to `messages.py`
 5. implement wrapper function in `comet.py`

**Comment by gavanderhoorn** [comment]
I'd gladly accept PRs adding support for new RPCs.

**Comment by rastreus** [comment]
> 1. figure out their names + (required) arguments (using Wireshark fi, or by looking at the Javascript)

- Can this be accomplished with a Virtual Robot (ROBOGUIDE, OlpcPro), or does it require a physical controller?
- I have little experience with Wireshark. What steps would properly determine an RPC API endpoint name and arguments?

**Comment by gavanderhoorn** [comment]
> > 1. figure out their names + (required) arguments (using Wireshark fi, or by looking at the Javascript)
> 
> * Can this be accomplished with a Virtual Robot (ROBOGUIDE, OlpcPro), or does it require a physical controller?

Yes, you can use Roboguide for this. The FRVC it uses essentially runs the same software.

> * I have little experience with Wireshark. What steps would properly determine an RPC API endpoint name and arguments?

All the RPCs I've seen so far use simple `GET` requests, so all arguments appear in the URL.

Example:

```
http://<CONTROLLER_IP>:80/COMET/rpc?func=REGVALRD&index=1
```

There might be some RPCs which take an actual JSON request document via `POST`, but I haven't encountered them. So it's mostly response documents which are JSON.

As to wireshark itself: you don't absolutely have to use it. You could use any HTTP-level traffic capture tool. Perhaps even your browser's *developer tools*.

Then browse to the main website of your robot's controller (ie: `http://localhost` if running a virtual controller in Roboguide), and open iRProgrammer (if you need to, in Roboguide, install `J767` on your virtual controller).

At that point start using the interface (ie: iRProgrammer) and keep an eye on your sniffer. You should see the requests as they are generated.

**Comment by gavanderhoorn** [comment]
I've just added an overview of "all" (?) RPCs and whether they're supported already or not. See [Supported RPCs](https://github.com/gavanderhoorn/comet_rpc/blob/master/README.md#supported-rpcs) in the readme.

**Comment by anandmit93** [comment]
Hello @gavanderhoorn ,

I tried inspecting the RPCs via developer tools and found some more of them. I have tested them for V9.30 and V9.40

IODRYRUN = "71" - Simulate all IOs
IOWETRUN = "72" - Unsimulate all IOs
XMLCOPY = "235" - Copy XML from one location to another 
Arguments:  from_path_name – Source file path; to_path_name – Destination file path
MNCPYPRG = "250" - Copy a TP program 
Arguments: src_name=source TP program name; chg_name=Copy of TP program name
SCDELETE = "15" - Delete lines in a TP program 
Arguments: prog_name= name of TP program; lin_num=Line number; &num=Number of lines to delete
MMDELPOS = "249" - Delete position data in a TP program (works only for Position P and not PR)
Arguments : prog_name= TP program name; pos_num= position index
SCEDIT = "29" - Add or replace line in a TP program 
Arguments: prog_name=TP program name; in_buf = text in url encoding; lin_num=linenumber to insert or replace; opt_sw=0/1/2 where 0 - overwrite 1 - Insert at that line, 2 - Insert after that line
SCSETPOS = "29" - Add or replace position data in a TP program Arguments: prog_name= TP program name; in_buf=position data in ASCII format 
TXLSTPRG_FC = "163" - List all programs 
Arguments: datatype=22 – types of acquired data – program name only/program name with comment..; prg_typ = 1 – tp/karel..; sub_typ=0 (unspecified/macro/condition/collection..)

MMDELPRG = "6" - Delete TP program 
Arguments: prog_name= Name of TP program
MMRENPRG = "116" - Rename a TP program 
Arguments: src_name = Name of TP program; chg_name = New name
MMSETATR = “21” – Set program attributes
Arguments: prog_name = Name of TP program ; attr = Attribute type ( comment /  write protection…); in_buf = value of the attribute
MMGETATR = “20” – Get program attributes
Arguments: prog_name = Name of TP program ; attr = Attribute type ( comment /  write protection…)

MNCHGREP = “243” – Change Position data representation
Arguments: in_buf = position data in ASCII format
MMCHGTYP = “23” – Change program subtype
Arguments: prog_name = Name of TP program ; sub_type =0 (unspecified/macro/condition/collection…)
REMARKLIN = “215” – Remark or unremark line(s)
Arguments: prog_name= Name of TP program; start = line number start; end = line number end; remark = 0/1 (unremark/remark)

**Comment by gavanderhoorn** [comment]
Hi, thanks.

I should perhaps have been clearer here: I believe we have all those RPCs in the readme:

https://github.com/gavanderhoorn/comet_rpc/blob/dcb6747bd60fbe2d61f97c1e4344db6594c10513/README.md?plain=1#L143-L229

What we need is for them to be supported by the library.

**Comment by anandmit93** [comment]
Sorry I forgot to mention the complete versions V9.30P/20 and V9.40P/35 & V9.40P/43. I guess only REMARKLIN seems to be missing in the list. Anyways, the support for these RPCs could be extended in this library.

**Comment by gavanderhoorn** [comment]
> Anyways, the support for these RPCs could be extended in this library.

agreed.

Using the existing implementation(s) as a template should make it not too difficult to get something working.

If you'd be up for it, I'd be more than willing to review some PRs getting support for some of those in.

Maybe start with `REMARKLIN`? That seems simple enough.

**Comment by anandmit93** [comment]
Sure, I will give it a try! I will update you after some progress has been made.

**Comment by gavanderhoorn** [comment]
From your description of the arguments, perhaps `paste_line` could be a good template:

https://github.com/gavanderhoorn/comet_rpc/blob/15f393960fc4055b3db4c663999d81eae15e7b72/src/comet_rpc/comet.py#L775-L817

**Comment by anandmit93** [comment]
Hello @gavanderhoorn ,

I have just raised a PR for 3 RPC functions. Please have a look at it.

**Comment by ghltrdi567** [comment]
> Hello @gavanderhoorn ,
> 
> I tried inspecting the RPCs via developer tools and found some more information about them. I have tested them for V9.30 and V9.40
> 
> ```
> IODRYRUN = "71" - Simulate all IOs
> IOWETRUN = "72" - Unsimulate all IOs
> XMLCOPY = "235" - Copy XML from one location to another 
> Arguments:  from_path_name – Source file path; to_path_name – Destination file path
> MNCPYPRG = "250" - Copy a TP program 
> Arguments: src_name=source TP program name; chg_name=Copy of TP program name
> SCDELETE = "15" - Delete lines in a TP program 
> Arguments: prog_name= name of TP program; lin_num=Line number; &num=Number of lines to delete
> MMDELPOS = "249" - Delete position data in a TP program (works only for Position P and not PR)
> Arguments : prog_name= TP program name; pos_num= position index
> SCEDIT = "29" - Add or replace line in a TP program 
> Arguments: prog_name=TP program name; in_buf = text in url encoding; lin_num=linenumber to insert or replace; opt_sw=0/1/2 where 0 - overwrite 1 - Insert at that line, 2 - Insert after that line
> SCSETPOS = "29" - Add or replace position data in a TP program Arguments: prog_name= TP program name; in_buf=position data in ASCII format 
> TXLSTPRG_FC = "163" - List all programs 
> Arguments: datatype=22 – types of acquired data – program name only/program name with comment..; prg_typ = 1 – tp/karel..; sub_typ=0 (unspecified/macro/condition/collection..)
> 
> MMDELPRG = "6" - Delete TP program 
> Arguments: prog_name= Name of TP program
> MMRENPRG = "116" - Rename a TP program 
> Arguments: src_name = Name of TP program; chg_name = New name
> MMSETATR = “21” – Set program attributes
> Arguments: prog_name = Name of TP program ; attr = Attribute type ( comment /  write protection…); in_buf = value of the attribute
> MMGETATR = “20” – Get program attributes
> Arguments: prog_name = Name of TP program ; attr = Attribute type ( comment /  write protection…)
> 
> MNCHGREP = “243” – Change Position data representation
> Arguments: in_buf = position data in ASCII format
> MMCHGTYP = “23” – Change program subtype
> Arguments: prog_name = Name of TP program ; sub_type =0 (unspecified/macro/condition/collection…)
> REMARKLIN = “215” – Remark or unremark line(s)
> Arguments: prog_name= Name of TP program; start = line number start; end = line number end; remark = 0/1 (unremark/remark)
> ```

Hello, @anandmit93 !

This is really great info.
I can repeat all neсessnary RPC`s exept SCSETPOS = "29" and MNCHGREP = “243”. 
Could you write valid  URL querry examples of this RPC?
Thank you in advance!

**Comment by gavanderhoorn** [comment]
Would be great to add some of these which are currently unsupported to the library @ghltrdi567.
